projects/plugins/jetpack/extensions/blocks/one-payment/edit.js

```diff
@@ -0,0 +1,103 @@
+import classnames from 'classnames';
+import { __ } from '@wordpress/i18n';
+import { useEffect } from '@wordpress/element';
+import { compose, withInstanceId } from '@wordpress/compose';
+import { createBlock, registerBlockVariation } from '@wordpress/blocks';
+import { withDispatch, withSelect } from '@wordpress/data';
+import {
+	InnerBlocks,
+	__experimentalBlockVariationPicker as BlockVariationPicker,
+} from '@wordpress/block-editor';
+
+/**
+ * Internal dependencies
+ */
+import defaultVariations from './variations';
+
+export function JetpackOnePaymentEdit( {
+	setAttributes,
+	hasInnerBlocks,
+	replaceBlock,
+	selectBlock,
+	clientId,
+	className,
+	blockType,
+	variations,
+	defaultVariation,
+} ) {
+	const formClassnames = classnames( className, 'jetpackone-payment' );
+
+	const createBlocksFromInnerBlocksTemplate = innerBlocksTemplate => {
+		const blocks = map( innerBlocksTemplate, ( [ name, attr, innerBlocks = [] ] ) =>
+			createBlock( name, attr, createBlocksFromInnerBlocksTemplate( innerBlocks ) )
+		);
+
+		return blocks;
+	};
+
+	const setVariation = variation => {
+		if ( variation.attributes ) {
+			setAttributes( variation.attributes );
+		}
+
+		if ( variation.innerBlocks ) {
+			replaceBlock( clientId, createBlocksFromInnerBlocksTemplate( variation.innerBlocks ) );
+		}
+
+		selectBlock( clientId );
+	};
+
+	useEffect( () => {
+		// Populate default variation on older versions of WP or GB that don't support variations.
+		if ( ! hasInnerBlocks && ! registerBlockVariation ) {
+			setVariation( defaultVariations[ 0 ] );
+		}
+	} );
+
+	const renderVariationPicker = () => {
+		return (
+			<div className={ formClassnames }>
+				<BlockVariationPicker
+					icon={ get( blockType, [ 'icon', 'src' ] ) }
+					label={ get( blockType, [ 'title' ] ) }
+					instructions={ __( "Please select which kind of payment you'd like to add.", 'jetpack' ) }
+					variations={ variations }
+					onSelect={ ( nextVariation = defaultVariation ) => {
+						setVariation( nextVariation );
+					} }
+				/>
+			</div>
+		);
+	};
+
+	if ( ! hasInnerBlocks && registerBlockVariation ) {
+		return renderVariationPicker();
+	}
+
+	return <InnerBlocks />;
+}
+
+export default compose( [
+	withSelect( ( select, props ) => {
+		const { getBlockType, getBlockVariations, getDefaultBlockVariation } = select( 'core/blocks' );
+		const { getBlocks } = select( 'core/block-editor' );
+		const innerBlocks = getBlocks( props.clientId );
+
+		return {
+			blockType: getBlockType && getBlockType( props.name ),
+			defaultVariation: getDefaultBlockVariation && getDefaultBlockVariation( props.name, 'block' ),
+			variations: getBlockVariations && getBlockVariations( props.name, 'block' ),
+			innerBlocks,
+			hasInnerBlocks: innerBlocks.length > 0,
+		};
+	} ),
+	withDispatch( dispatch => {
+		const { replaceBlock, selectBlock } = dispatch( 'core/block-editor' );
+		return { replaceBlock, selectBlock };
+	} ),
+	withInstanceId,
+] )( JetpackOnePaymentEdit );
```

Comment:
I think we should go with hooks rather than compose to keep the code more modern. 🙂 

E.g.
```js
import { useDispatch, useSelect } from '@wordpress/data';

export default function JetpackOnePaymentEdit( {
	name,
	setAttributes,
	clientId,
	className,
} ) {
	const {
		blockType,
		defaultVariation,
		variations,
		hasInnerBlocks
	} = useSelect( select => {
		const { getBlockType, getBlockVariations, getDefaultBlockVariation } = select( 'core/blocks' );
		const { getBlocks } = select( 'core/block-editor' );

		return {
			blockType: getBlockType( name ),
			defaultVariation: getDefaultBlockVariation( name, 'block' ),
			variations: getBlockVariations( name, 'block' ),
			hasInnerBlocks: getBlocks( clientId )?.length > 0,
		};
	} );
	const { replaceBlock, selectBlock } = useDispatch( 'core/block-editor' );
	/* ... */
}
```

In my example, I've also removed `withInstanceId`, that AFAICS is not needed (nor used) here.

Reply:
Oh, also a minor detail that I've just remembered: in the last few months, Core Gutenberg has started using imports instead of hardcoded store names.

Here it would look like this:

```js
import { /* ... */, store as blocksStore } from '@wordpress/blocks';
import { /* ... */, store as blockEditorStore } from '@wordpress/block-editor';

/* ... */

select( blocksStore );
select( blockEditorStore );

/* ... */

useDispatch( blockEditorStore );
```